homeassistant/components/homekit_controller/utils.py

```diff
@@ -0,0 +1,40 @@
+"""Helper functions for the homekit_controller component."""
+from typing import cast
+
+from aiohomekit import Controller
+
+from homeassistant.components import zeroconf
+from homeassistant.const import EVENT_HOMEASSISTANT_STOP
+from homeassistant.core import Event, HomeAssistant
+
+from .const import CONTROLLER
+
+
+async def async_get_controller(hass: HomeAssistant) -> Controller:
+    """Get or create an aiohomekit Controller instance."""
+    if existing := hass.data.get(CONTROLLER):
+        return cast(Controller, existing)
+
+    async_zeroconf_instance = await zeroconf.async_get_async_instance(hass)
+
+    # In theory another call to async_get_controller could have run while we were
+    # trying to get the zeroconf instance. So we check again to make sure we
+    # don't leak a Controller instance here.
+    if existing := hass.data.get(CONTROLLER):
+        return cast(Controller, existing)
+
+    controller = Controller(async_zeroconf_instance=async_zeroconf_instance)
+
+    hass.data[CONTROLLER] = controller
+
+    async def _async_stop_homekit_controller(event: Event) -> None:
+        # Pop first so that in theory another controller /could/ start
+        # While this one was shutting down
+        hass.data.pop(CONTROLLER, None)
+        await controller.async_stop()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, _async_stop_homekit_controller)
```

Comment:
Does this need to be canceled when the last homekit_controller config entry is unloaded?

Reply:
One can get started during a step_user config flow as well, so it would be when HA was stopped, when homekit_controller was unloaded and when there were no active homekit_controller config flows.

Reply:
If the controller is ever stopped outside of the EVENT_HOMEASSISTANT_STOP event, it will to leak a listener since if you don't call the callable that `async_listen_once` returns which removes the listener from the underlying dict.

Reply:
I don't think you ever stop it outside of the stop event though so a comment that its never stopped until the stop event so there is no need to cancel the listener would be fine as well

Reply:
Added comment in 0664c8bbe892d781b9716238d6de24aac352d0a3.